### PR TITLE
Update EMC dev/emc from NCAR dtc/develop 2019/12/17

### PR DESCRIPTION
**Pull request by climbfuji**
This PR contains:
- add new suite definition files for GFSv15p2, GFSv16beta
- update existing suite definition files because of the interstitial code rearrangement (https://github.com/NCAR/ccpp-physics/pull/372)
- clean up CCPP compiler flags, add -Wall for GNU in DEBUG mode
- add satmedmfvifq, shalcnv, sascnvn, Ferrier-Aligo microphysics and dependencies to CCPP prebuild config
- bug fixes for ugwp and noahmp
- move previous-timestep precipitation variables from Tbd to Sfcprop (as recommended by @HelinWei-NOAA, see NCAR/ccpp-physics#319)
- fix compiler warnings about non-existent include directories
- cleanup of old comments in GFS_physics_driver.F90 (see NCAR#4, NCAR#16, and NCAR/ccpp-physics#319
- CCPP annotations in `GFS_driver.F90`, `GFS_radiation_driver.F90`, `GFS_physics_driver.F90` (comments that describe where code blocks ended up in the CCPP interstitial code)
- capability to coldstart FV3 with GSD physics from RAP/HRRR initial conditions (required for FV3 SAR)
- new suite definition file for coupled model

Associated PRs:
https://github.com/NOAA-EMC/GFDL_atmos_cubed_sphere/pull/11
https://github.com/NCAR/ccpp-framework/pull/246
https://github.com/NCAR/ccpp-physics/pull/372
https://github.com/NOAA-EMC/fv3atm/pull/30
https://github.com/ufs-community/ufs-weather-model/pull/18

For regression testing, see https://github.com/ufs-community/ufs-weather-model/pull/18.

### CODE REVIEW

**Review by HelinWei-NOAA — CHANGES_REQUESTED** [review]
Since both Noah MP and RUC use the same variables for different precipitation types, we should save them for restart for both models. So in GFS_restart.F90, we need to add the case for Noah MP. For example,

! RUC 
    if (Model%lsm == Model%lsm_ruc) then
      Restart%num2d = Restart%num2d + 5
    endif

Changed to

! RUC and Noah MP
    if (Model%lsm == Model%lsm_ruc. or. Model%lsm == Model%lsm_noahmp) then
      Restart%num2d = Restart%num2d + 5
    endif

**Comment by HelinWei-NOAA** [comment]
For all those extra variables added to DDT sfcprop, we didn't add how to cold-start them in FV3GFS_io.F90. Also for NoahMP, we may not need to have both precip amount and precip rate, we can convert them inside land surface driver. In this way the structure looks more like RUC.

**Comment by HelinWei-NOAA** [comment]
There is another issue with Noah MP. I am not sure if we should fix with this commit or later. 
If Noah MP is running without any microphysics scheme, there will be no any precip forcing. In this case we have to use the total precip and partition it into rain or snow based on temperature.

**Comment by climbfuji** [comment]
> For all those extra variables added to DDT sfcprop, we didn't add how to cold-start them in FV3GFS_io.F90. Also for NoahMP, we may not need to have both precip amount and precip rate, we can convert them inside land surface driver. In this way the structure looks more like RUC.

Understand. Enabling b4b identical restart runs for NoahMP is not in the scope of this PR, neither is rewriting the logic inside NoahMP. We only wanted to fix the bug that caused using zero precipitation rates at fhzero intervals. If there is an opportunity to streamline the logic used by both RUC LSM and NoahMP that also avoids writing 10 instead of 5 2d arrays to the restart files for b4b identical restart runs using NoahMP, then this will make life a lot easier in the future. But I think this work should be done by your team (we will be there to help with the CCPP part if needed) in a follow-up change.

**Comment by climbfuji** [comment]
> There is another issue with Noah MP. I am not sure if we should fix with this commit or later.
> If Noah MP is running without any microphysics scheme, there will be no any precip forcing. In this case we have to use the total precip and partition it into rain or snow based on temperature.

Good point, I think this should go together with the work that you described just before regarding streamlining the logic used for NoahMP and RUC.

**Comment by HelinWei-NOAA** [comment]
Dom, with the current modification, running fv3gfs with noah mp option can't be restarted properly. So we may use the easy fix for Noah MP precip forcing bug suggested by Moorthi: 

comment out the following lines in subroutine diag_phys_zero
    Diag%rain       = zero
    Diag%rainc      = zero
    Diag%ice        = zero
    Diag%snow       = zero
    Diag%graupel    = zero

These five fields are not accumulated and no need zero out in diagnostic bucket.

**Comment by climbfuji** [comment]
> Dom, with the current modification, running fv3gfs with noah mp option can't be restarted properly. So we may use the easy fix for Noah MP precip forcing bug suggested by Moorthi:
> 
> comment out the following lines in subroutine diag_phys_zero
> Diag%rain = zero
> Diag%rainc = zero
> Diag%ice = zero
> Diag%snow = zero
> Diag%graupel = zero
> 
> These five fields are not accumulated and no need zero out in diagnostic bucket.

I would like @grantfirl's opinion on this one as well. I believe that without/before our modifications, NoahMP couldn't be restarted properly as well? Is it the purpose of this PR to fix it?

**Comment by HelinWei-NOAA** [comment]
Dom,

Before your modification, Noah MP can be restarted properly because those
five precip variables in DDT Diag are stored in restart files. After your
modification, Noah MP will use the data from DDT Sfcprop. However they are
not stored in restart file for Noah MP.

Helin

**Comment by climbfuji** [comment]
> Dom, Before your modification, Noah MP can be restarted properly because those five precip variables in DDT Diag are stored in restart files. After your modification, Noah MP will use the data from DDT Sfcprop. However they are not stored in restart file for Noah MP. Helin
> […](#)

I am sorry but I am confused. Where are the five diag variables written to the restart files? If you check GFS_restart.F90, then you see that the five variables that were previously in Tbd and now in Sfcprop are written to the restart files. I checked FV3GFS_io.F90 and GFS_restart.F90, and I don't see any place where IntDiag fields are written to restart files. I also checked the restart output in the EMC baseline for the NoahMP test, `/scratch1/NCEPDEV/nems/emc.nemspara/RT/NEMSfv3gfs/trunk-20191127/fv3_gfdlmprad_noahmp/RESTART`. I don't see any file containing those five precipitation fields. I must be missing something. Can you please point me to the lines of code that are writing these IntDiag fields to disk in the current develop branch of fv3atm? Thanks!

**Comment by HelinWei-NOAA** [comment]
I think you are right. We have the restarting issue for Noah MP before and
after your modifications. Noah MP with FV3GFS still has some other issues
need to be fixed. Therefore we should have another PR exclusively to
address all remaining issues for Noah MP. I will recommend Jun to go ahead
and commit your changes. Thanks.

**Comment by climbfuji** [comment]
Thanks for checking, Helin. It's always good to have more eyes looking at the code.

**Review by HelinWei-NOAA — APPROVED** [review]

**Comment by Qingfu-Liu** [comment]
I only go through two directories which are physics related, and compared the contents between Dom's fork and the current UFS repository. The directories are:
ufs-weather-model-DOM/FV3/gfsphysics
ufs-weather-model-DOM/FV3/ccpp
There are few changes in the IPD driver and its related physics schemes. All the changes look fine to me. The major changes are CCPP related. There are several new schemes (new suites) in Dom's fork. These new schemes (new suites) can only be called using CCPP, not IPD. In Dom's fork, some of the post_runs are merged into GFS_GWD_generic_post, so developers need to update their SDF if they have not committed their suite to the CCPP repository. All the changes look fine to me, but I can't say much to the changes of the physics schemes which I am not familiar with (for example, sfc_drv_ruc.F90). Since all the regression tests are successful,  I think it is fine to merge Dom's fork to the EMC repository.

File: gfsphysics/GFS_layer/GFS_physics_driver.F90 [hunk 1]
```diff
@@ -3698,7 +3869,7 @@ subroutine GFS_physics_driver                                 &
 
           else      ! ras version 2
 
-! DH* this code not yet in CCPP (belongs to GFS_DCNV_generic_pre?)
+!## CCPP ##* This code not yet in CCPP Note: Likely belongs in rascnv_pre.
```
Comment (SMoorthi-emc) [inline gfsphysics/GFS_layer/GFS_physics_driver.F90:3872]: I have moved this calculation inside RAS itself.  There is no need for rascnv_pre

File: gfsphysics/GFS_layer/GFS_physics_driver.F90 [hunk 2]
```diff
@@ -3730,14 +3901,14 @@ subroutine GFS_physics_driver                                 &
...
+!*## CCPP ##
 !           if (lprnt) write(0,*)' gt04bras=',Stateout%gt0(ipr,:)
 !           if (lprnt) write(0,*)' gq04bras=',Stateout%gq0(ipr,:,1)
 !           if (lprnt) write(0,*)'befrasclw1=',clw(ipr,:,1)
 !           if (lprnt) write(0,*)'befrasclw2=',clw(ipr,:,2)
 !           if (lprnt) write(0,*)'befrastke=',clw(ipr,:,ntk)
 !           if (lprnt) write(0,*)'trcmin=',trcmin(ntk-2),' ntk=',ntk
-
+!## CCPP ## Not in CCPP yet.
             call rascnv (im, ix, levs, dtp, dtf, Tbd%rann, Stateout%gt0,     &
```
Comment (SMoorthi-emc) [inline gfsphysics/GFS_layer/GFS_physics_driver.F90:3912]: it is now in my ccpp fork.

**Review by Qingfu-Liu — APPROVED** [review]
I only go through two directories which are physics related, and compared the contents between Dom's fork and the current UFS repository. The directories are:
ufs-weather-model-DOM/FV3/gfsphysics
ufs-weather-model-DOM/FV3/ccpp
There are few changes in the IPD driver and its related physics schemes. All the changes look fine to me. The major changes are CCPP related. There are several new schemes (new suites) in Dom's fork. These new schemes (new suites) can only be called using CCPP, not through IPD driver. In Dom's fork, some of the post_runs are merged into GFS_GWD_generic_post, so developers need to update their SDF if they have not committed their suite to the CCPP repository. All the changes look fine to me, but I can't say much to the changes of the physics schemes which I am not familiar with (for example, sfc_drv_ruc.F90). Since all the regression tests are successful, I think it is fine to merge Dom's fork to the EMC repository.

**Review by SMoorthi-emc — APPROVED** [review]
Looks ok to me.

**Comment by climbfuji** [comment]
> Looks ok to me.

Thanks, Moorthi. I suggest to update the comments in GFS_physics_driver.F90 when your RAS-CCPP work is merged. OK?

**Review by junwang-noaa — APPROVED** [review]
The code changes look good to me.

**Review by DusanJovic-NOAA — APPROVED** [review]
Approved, based on other people's reviews.

**Pull request merged**